koswat/strategies/infra_priority_strategy/infra_priority_strategy.py

```diff
@@ -65,44 +78,107 @@ def _get_common_available_measures(
             if _cd_m in _common_available_measures
         }
 
-    def _set_cheapest_common_available_measure(
-        self,
-        initial_measure: type[ReinforcementProfileProtocol],
-        cluster_data: list[StrategyLocationReinforcement],
-        measure_costs: dict[type[ReinforcementProfileProtocol]],
-    ):
-        # 3.1. Get the cheapest common available measure.
-        _selection = min(
-            measure_costs,
-            key=measure_costs.get,
-        )
+    @staticmethod
+    def generate_subcluster_options(
+        from_cluster: InfraCluster, min_length: int
+    ) -> Iterator[list[InfraCluster]]:
+        """
+        Generates all possible combinations of (sub) clusters based on the locations
+        of this cluster. These are also referred as "options".
+
+        Args:
+            min_length (int): Minimun length for a cluster to be valid.
 
-        # 3.2. Set the selection if different from the initial.
-        if _selection != initial_measure:
-            for _cd in cluster_data:
-                _cd.selected_measure = _selection
+        Returns:
+            list[list[InfraCluster]]: Collection of "options" with valid length.
+        """
+
+        if not from_cluster.fits_subclusters():
+            # If the cluster is not twice as big as the minimum
+            # required length we know there is no possibility of
+            # creating subclusters.
+            return [[from_cluster]]
+
+        def get_cluster_option(
+            location_collection: list[StrategyLocationReinforcement],
+        ) -> Iterator[list[InfraCluster]]:
+            _icc = []
+            for _w_element in filter(len, location_collection):
+                _ic = InfraCluster(
+                    min_required_length=from_cluster.min_required_length,
+                    reinforcement_type=from_cluster.reinforcement_type,
+                    cluster=list(_w_element),
+                )
+
+                if not _ic.is_valid():
+                    # There is no need to check further,
+                    # this option will result as not valid.
+                    _icc.clear()
+                    break
+                _icc.append(_ic)
+            return _icc
+
+        return filter(
+            len,
```

Comment:
It's not an undocumented feature, filter will execute the callable (first argument) to consume the collection (second argument) and will return the items that evaluate to true.
- `len` is a built-in operator which returns the length of a collection,
- Since Python 3 (I think), thus an empty list will evaluate to false whilst with values to true.

All being said, I found this approach online and seems pretty neat as it avoids instantiating directly.